Fixes to Music Playlist Editor

DaveTBlake opened the pull request:
A few more fixes to the music playlist editor screen 

 - Sources list (left panel) context menu
Ensure the only context menu button for items on the left hand side sources list is "add" .
Despite being a descendent of  `CGUIWindowMusicBase` this screen does not need the standard media window context menu buttons on items of either of the list controls. Having navigated within one of the top system items ("Files" / "Library" / "Music Videos") of the left hand list the only action is to add the music files of the item to the playlist.
- Reinstate U/D/Del keyboard shortcuts 
Keyboard controls U, D and Del  can again be used in the right hand panel to move the highlighted item Up, Down or Deleted. Lost in the previous fix #18838.
- Remove "Add music source" item from within "files" source list
- On click behaviour on items in sources list 
Browse into .m3u files (treat like folder), avoid starting playback of media files (do nothing)
- Open button shows music playlists
- Save notification
Add a "saved" toast when playlist saved on clicking  save button.
- Music Videos
Correct URL (not virtual path) saved to .m3u file, so playback of the resulting playlists does not skip those videos

DaveTBlake committed:
Fixes for music playlist editor
- reinstate u/d/del keyboard shortcuts
- context menu for items on sources list to be "add" only.
- remove "add music source" item from within "files" source list
- "library" shows list of users music library nodes (like main music library screen)
- on click on items in sources list browse into .m3u files (treat like folder), avoid starting playback of media files (do nothing)
- open button shows music playlists
- add notification when playlist saved
- save correct URL for music videos (not virtual path)

DaveTBlake commented:
Win64 test build including this change http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20201124-fa0fd4c6-FixesForBeta2-x64.exe

KarellenX commented:
I tested the basic playlist editor:
- the context menu in the right panel works correctly with the correct options
- the Files and Library selection in the middle panel point to the correct locations and the context menu shows Add as the only selection

DaveTBlake commented:
> In the right panel, using keyboard controls no longer works. Using U, D and Del would move the highlighted item Up, Down or Deleted. 

Ah, that would probably have gone with the previous fix https://github.com/xbmc/xbmc/pull/18838, not touched by this PR. But I will see if I can restore it.

jjd-uk commented:
> Also, as a non-related issue, when pressing Save there is no feedback. A couple of times I sat there for a few seconds waiting for something to happen until I remembered nothing happens. Maybe a pop-up stating "Playlist saved" or something, especially as we remain in the Playlist screen, and not taken back to the playlist node.

I agree, it would be nice to have a Dialog Notification to say it's been saved.

DaveTBlake commented:
Have added extra fixes as requested, a couple of things  I spotted along the way  and discussed on Slack (and updated PR description).
- Reinstated U/D/Del keyboard shortcuts 
- Add a "saved" toast when playlist saved on clicking  save button
- Removed "Add music source" item from within "files" source list
- Browse into .m3u files (treat like folder) when click on item in sources list
- Avoid starting playback of media files (do nothing) when click on item in sources list.
- Open button shows contents of music playlists folder.
This is where save stores the edited playlist, any other playlist files located elsewhere on music source scan be accessed via the left hand panel of sources and added to the list   being edited and thus saved in the music playlists folder.

KarellenX commented:
>Sources list (left panel) context menu
>Ensure the only context menu button for items on the left hand side sources list is "add" .
>Despite being a descendent of CGUIWindowMusicBase this screen does not need the standard media window context menu buttons on items of either of the list controls. Having navigated within one of the top system items ("Files" / "Library" / "Music Videos") of the left hand list the only action is to add the music files of the item to the playlist.

Confirmed. There is only an **Add** option in the context menu. I can Add music at the Source, Album and Song levels depending on folder structure. If I place a m3u playlist in my Source, maybe one that I created elsewhere, I can use the Add option and copy the m3u playlist songs into the current playlist being edited.

>Reinstate U/D/Del keyboard shortcuts
>Keyboard controls U, D and Del can again be used in the right hand panel to move the highlighted item Up, Down or Deleted. Lost in the previous fix #18838.

U/D/Del shortcuts are again working in the right panel.

>Remove "Add music source" item from within "files" source list

Yes that is gone. To now add songs to the playlist, a Source must be added first from the normal location.

>On click behaviour on items in sources list
>Browse into .m3u files (treat like folder), avoid starting playback of media files (do nothing)

Not sure what to do here. Do I need the `<playlistasfolder>` tag for this?

>Open button shows music playlists

Yes it shows the m3u playlists in the /playlists/music/ folder.

>Save notification
>Add a "saved" toast when playlist saved on clicking save button.

Yes, a Playlist Saved toast pops out.

>Music Videos
>Correct URL (not virtual path) saved to .m3u file, so playback of the resulting playlists does not skip those videos

The file path is saved to the playlist. Music Videos and Music now play in the correct order but...
1. If the first item is a Music Video followed by a song, Kodi continues to use VideoPlayer for songs and we end up with a black screen.
2. If the first item is a song followed by a music video, then Kodi continues to use PAPlayer for the music video and we get glitchy audio with fanart remaining on the screen. Then after the song ends Kodi plays up and no more audio.

DaveTBlake commented:
Thanks for retesting @KarellenX. The problems you report are not related to this PR or the music playlist editor, but obviously since the editor can now add music videos to the .m3u file3 correctly are easier to reproduce. More work for another PR.

Win64 test build for this PR here  http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20201126-a4b3778f-PlaylistEditorFix2-x64.exe is anyone else wants to try it.

DaveTBlake commented:
jenkins build this please

jjd-uk commented:
The ability to access the Playlists folder in the middle section in order to use the Add option is no longer possible, there's no Playlists option available under either Files (where it was previously) or Library.  This is needed to give the ability to merge playlists.

Previously:

1. Select new playlist
2 . Select Files the Playlists
3. Select m3u and select Add from context menu
4. Select a different m3u and select Add from context menu
5. Save with new playlist name.

![image](https://user-images.githubusercontent.com/5781142/100473275-dedb1880-30d5-11eb-9a24-36f735eb7420.png)

![image](https://user-images.githubusercontent.com/5781142/100473290-e7cbea00-30d5-11eb-941a-3510832aa541.png)

DaveTBlake commented:
Access to the Playlists folder on the left hand panel was an accidient, aviallable only because the editor was broken and a regression was causing "Files" to show a listing of the users root music menu instead of music sources. If you go back to v15.2 (last version that the editor was fully working)  it does NOT have "playlists" under that panel. 

However I agree that it is useful to be able to pick music from one playlist (or smart playlist) in the Playlists folder and add it to the playlist you are currently editing.  After discussion on Slack, this is best provided by changing "Library" to lead to a listing of the users root music menu which includes the Playlists folder. Historically "Library" had path "musicdb://" and so only showed the original directory nodes.

I have updated this PR accordingly

DaveTBlake commented:
Win64 test build of this + partymode playlist context menu fixes here http://mirrors.kodi.tv/test-builds/windows/win64/KodiSetup-20201128-d2c0fe85-Playcontext1-x64.exe

KarellenX commented:
The Library option in the centre panel now shows "Playlists". Adding either the full playlist or individual songs from the playlist works as expected when using either keyboard commands or context menu.

Smart Playlists and "New..." options are shown also. Neither of these can be added to the playlist.

All seems to be working ok. Thanks @DaveTBlake !!

jjd-uk commented:
A bit late to this one, I can confirm that the above playlist test case is now fixed with Library option now pointing to `library://music/`. Thanks Dave for the effort in sorting out this neglected section of Kodi.

DaveTBlake commented:
I took the +1 as good to go @jjd-uk and so did. Thanks for discssing and testing.
